coast/gridded.py

```diff
@@ -0,0 +1,865 @@
+        H = e3t_0.cumsum(dim='z_dim').isel(z_dim=ds_dom.bottom_level.astype("int") - 1)
+        # Add correction to e3t_0 due to change in ssh
+        e3t_new = e3t_0 * (1 + ssh / H)
+        # preserve dimension ordering
+        e3t_new = e3t_new.transpose('t_dim', 'z_dim', 'y_dim', 'x_dim')
+        # mask out correction at layers below bottom level
+        e3t_new = e3t_new.where(e3t_new.z_dim < ds_dom.bottom_level, e3t_0.data)
+        # preserve any other t mask
+        e3t_new = e3t_new.where(~np.isnan(ssh))
+        if e3t:
+            e3_return.append(e3t_new.squeeze())
+
+        if np.any([e3u, e3v, e3f]):
+            e1e2t = ds_dom.e1t * ds_dom.e2t
+        if np.any([e3u, e3v, e3w]):
+            e3t_dt = e3t_new - e3t_0
+
+        # area averaged interpolation onto the u-grid to get e3u    
+        if np.any([e3u, e3f]):
+            e1e2u = ds_dom.e1u * ds_dom.e2u
+            # interpolate onto u-grid
+            e3u_temp = (((0.5 / e1e2u[:, :-1]) * ((e1e2t[:, :-1] * e3t_dt[:, :, :, :-1])
+                                                  + (e1e2t[:, 1:] * e3t_dt[:, :, :, 1:])))
+                        .transpose('t_dim', 'z_dim', 'y_dim', 'x_dim'))
+            # u mask
+            e3u_temp = e3u_temp.where(e3t_dt[:, :, :, 1:] != 0, 0)
+            # mask out correction at layers below bottom level
+            e3u_temp = e3u_temp.where(e3u_temp.z_dim < ds_dom.bottom_level[:, :-1], 0)
+            # Add correction to e3u_0
+            e3u_temp = (e3u_temp + ds_dom.e3u_0[:, :, :-1])
+            e3u_new = xr.zeros_like(e3t_new)
+            e3u_new[:, :, :, :-1] = e3u_temp
+            e3u_new[:, :, :, -1] = ds_dom.e3u_0[:, :, -1]
+            e3u_new['longitude'] = ds_dom.glamu
+            e3u_new['latitude'] = ds_dom.gphiu
+            if e3u:
+                e3_return.append(e3u_new.squeeze())
+
+        # area averaged interpolation onto the u-grid to get e3v 
+        if e3v:
+            e1e2v = ds_dom.e1v * ds_dom.e2v
+            e3v_temp = (((0.5 / e1e2v[:-1, :]) * ((e1e2t[:-1, :] * e3t_dt[:, :, :-1, :])
+                                                  + (e1e2t[1:, :] * e3t_dt[:, :, 1:, :])))
+                        .transpose('t_dim', 'z_dim', 'y_dim', 'x_dim'))
+            e3v_temp = e3v_temp.where(e3t_dt[:, :, 1:, :] != 0, 0)
+            e3v_temp = e3v_temp.where(e3v_temp.z_dim < ds_dom.bottom_level[:-1, :], 0)
+            e3v_temp = (e3v_temp + ds_dom.e3v_0[:, :-1, :])
+            e3v_new = xr.zeros_like(e3t_new)
+            e3v_new[:, :, :-1, :] = e3v_temp
+            e3v_new[:, :, -1, :] = ds_dom.e3v_0[:, -1, :]
+            e3v_new['longitude'] = ds_dom.glamv
+            e3v_new['latitude'] = ds_dom.gphiv
+            e3_return.append(e3v_new.squeeze())
+
+        # area averaged interpolation onto the u-grid to get e3f
+        if e3f:
+            e1e2f = ds_dom.e1f * ds_dom.e2f
+            e3u_dt = e3u_new - ds_dom.e3u_0
```

Comment:
Similar to the previous comment, we need to either ensure that e3u_new and e1e2u are set or or give a "friendly" error.